### PR TITLE
Update traits-based games

### PR DESCRIPTION
**Pull request by szhorvat**
This PR makes the changes detailed in #1620 .

Details:

 - Add `const` to input-only arguments.
 - Improve docs (e.g. `edges_per_step` is not the number of edges added, but the number of edges tried to be added with some probability)
 - callaway_traits and establishment now return vertex types in a new output argument
 - callaway_traits and establishment: `type_dist` is now allowed to be `NULL` in which case it is taken to be uniform
 - improve parameter checks
 - asymmetic_preference: allow different number of in- and out-types, reorder argument to always be out-types, then in-types as this feel natural given the interpretation of the type distribution matrix and preference matrix (rows: out, cols: in).
 - update old tests from `return`-style to `ASSERT`-style (otherwise the leak detector complains on failure)

### CODE REVIEW

**Comment by szhorvat** [comment]
@GroteGnoom : I'm planning to touch `establishment_game`, so please wait with tests for that function. If you're already working on it,  let me know.

**Comment by GroteGnoom** [comment]
> @GroteGnoom : I'm planning to touch `establishment_game`, so please wait with tests for that function. If you're already working on it, let me know.

I've skipped it for now, thanks for asking. Go ahead!

File: src/games/establishment.c [hunk 1]
```diff
@@ -54,42 +58,100 @@
...
+
+    if (! directed && ! igraph_matrix_is_symmetric(pref_matrix)) {
+        IGRAPH_ERROR("The preference matrix must be symmetric when generating undirected graphs.", IGRAPH_EINVAL);
+    }
+
+    {
+        igraph_real_t lo, hi;
+        igraph_matrix_minmax(pref_matrix, &lo, &hi);
+
+        if (lo < 0 || hi > 1) {
```
Comment (ntamas) [inline src/games/establishment.c:104]: What if a NaN somehow sneaks into the preference matrix? Both the comparisons will be `false` if I understand correctly and we won't hit this branch, right?

**Review by ntamas — APPROVED** [review]
Looks good to me!

File: src/games/establishment.c (hunk 1 above)
Comment (szhorvat) [reply src/games/establishment.c:104]: OK, will update all of them to detect NaNs. I originally started this PR before we added NaN detection.

Comment (szhorvat) [reply src/games/establishment.c:104]: Next we have to add some helpers which do the NaN check and error reporting automatically. Writing the same code again and again is getting tedious.

Comment (ntamas) [reply src/games/establishment.c:104]: A quick (although awkward) fix is to reverse the comparison: `!(lo >= 0 && hi <= 1)`

Comment (ntamas) [reply src/games/establishment.c:104]: Feel free to merge when you have updated the PR.

**Review by ntamas — APPROVED** [review]

**Pull request merged**